kernel/sched.c

```diff
@@ -489,20 +491,38 @@ void z_thread_single_abort(struct k_thread *thread)
 		}
 
 		thread->base.thread_state |= mask;
-	}
-
-	sys_trace_thread_abort(thread);
 
 #ifdef CONFIG_USERSPACE
-	/* Clear initialized state so that this thread object may be re-used
-	 * and triggers errors if API calls are made on it from user threads
-	 */
-	z_object_uninit(thread->stack_obj);
-	z_object_uninit(thread);
+		/* Clear initialized state so that this thread object may be
+		 * re-used and triggers errors if API calls are made on it from
+		 * user threads
+		 *
+		 * For a whole host of reasons this is not ideal and should be
+		 * iterated on.
+		 */
+		z_object_uninit(thread->stack_obj);
+		z_object_uninit(thread);
 
-	/* Revoke permissions on thread's ID so that it may be recycled */
-	z_thread_perms_all_clear(thread);
+		/* Revoke permissions on thread's ID so that it may be
+		 * recycled
+		 */
+		z_thread_perms_all_clear(thread);
 #endif
+
+		/* Wake everybody up who was trying to join with this thread.
+		 * A reschedule is invoked later by k_thread_abort().
+		 */
+		while ((waiter = z_waitq_head(&thread->base.join_waiters)) !=
+		       NULL) {
+			_priq_wait_remove(&pended_on(waiter)->waitq, waiter);
```

Comment:
This sequence needs to be locked on SMP, both because the wait queue operation can race with other threads/cpus and because other contexts may be trying to wait on the same thread we're destroying here and try to modify the join_waiters list.

Reply:
this code is inside a LOCKED(&sched_spinlock). Is there some other lock I need to be holding?

Reply:
Ah, OK.  Missed the extra level of indentation that moved the end of the LOCKED() block.